pgrx/src/spi/tuple.rs

```diff
@@ -0,0 +1,529 @@
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if somehow the specified name contains a null byte.
+    pub fn get_datum_by_name<S: AsRef<str>>(
+        &self,
+        name: S,
+    ) -> std::result::Result<&SpiHeapTupleDataEntry<'conn>, Error> {
+        unsafe {
+            let name_cstr = CString::new(name.as_ref()).expect("name contained a null byte");
+            let fnumber = pg_sys::SPI_fnumber(self.tupdesc.as_ptr(), name_cstr.as_ptr());
+
+            if fnumber == pg_sys::SPI_ERROR_NOATTRIBUTE {
+                Err(Error::SpiError(SpiErrorCodes::NoAttribute))
+            } else {
+                self.get_datum_by_ordinal(fnumber as usize)
+            }
+        }
+    }
+
+    /// Set a datum value for the specified ordinal position
+    ///
+    /// # Errors
+    ///
+    /// If the specified ordinal is out of bounds a [`SpiErrorCodes::NoAttribute`] is returned
+    pub fn set_by_ordinal<T: IntoDatum>(
+        &mut self,
+        ordinal: usize,
+        datum: T,
+    ) -> std::result::Result<(), Error> {
+        self.check_ordinal_bounds(ordinal)?;
+        self.entries[ordinal - 1] = SpiHeapTupleDataEntry {
+            datum: datum.into_datum(),
+            type_oid: T::type_oid(),
+            __marker: PhantomData,
+        };
+        Ok(())
+    }
+
+    /// Set a datum value for the specified field name
+    ///
+    /// # Errors
+    ///
+    /// If the specified name isn't valid a [`Error::SpiError(SpiError::NoAttribute)`] is returned
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if somehow the specified name contains a null byte.
+    pub fn set_by_name<T: IntoDatum>(
+        &mut self,
+        name: &str,
+        datum: T,
+    ) -> std::result::Result<(), Error> {
+        unsafe {
+            let name_cstr = CString::new(name).expect("name contained a null byte");
+            let fnumber = pg_sys::SPI_fnumber(self.tupdesc.as_ptr(), name_cstr.as_ptr());
+            if fnumber == pg_sys::SPI_ERROR_NOATTRIBUTE {
+                Err(Error::SpiError(SpiErrorCodes::NoAttribute))
+            } else {
+                self.set_by_ordinal(fnumber as usize, datum)
+            }
+        }
+    }
+
+    #[inline]
+    pub fn columns(&self) -> usize {
+        unsafe {
+            // SAFETY: we know self.tupdesc is a valid, non-null pointer because we own it
+            (*self.tupdesc.as_ptr()).natts as usize
+        }
+    }
+
+    /// is the specified ordinal valid for the underlying tuple descriptor?
+    #[inline]
+    fn check_ordinal_bounds(&self, ordinal: usize) -> std::result::Result<(), Error> {
+        if ordinal < 1 || ordinal > self.columns() {
+            Err(Error::SpiError(SpiErrorCodes::NoAttribute))
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl<'conn> SpiHeapTupleDataEntry<'conn> {
+    pub fn value<T: IntoDatum + FromDatum>(&self) -> SpiResult<Option<T>> {
```

Comment:
Part of the problem really does go down all the way here, where it becomes possible to extract types with only FromDatum as the effective boundary on them.